crates/lib/kajiya-egui/src/egui_backend.rs

```diff
@@ -0,0 +1,298 @@
+    pub egui_texture: Arc<Image>,
+}
+
+pub struct EguiBackendInner {
+    egui_renderer: ash_egui::Renderer,
+    gfx: Option<GfxResources>,
+}
+
+pub struct EguiBackend {
+    inner: Arc<Mutex<EguiBackendInner>>,
+    device: Arc<Device>,
+    pub raw_input: ash_egui::egui::RawInput,
+}
+
+impl EguiBackend {
+    pub fn new(
+        device: Arc<Device>,
+        window_settings: (u32, u32, f64),
+        context: &mut CtxRef,
+    ) -> Self {
+        let (window_width, window_height, window_scale_factor) = window_settings;
+
+        // Create raw_input
+        let raw_input = egui::RawInput {
+            pixels_per_point: Some(window_scale_factor as f32),
+            screen_rect: Some(egui::Rect::from_min_size(
+                Default::default(),
+                vec2(window_width as f32, window_height as f32) / window_scale_factor as f32,
+            )),
+            time: Some(0.0),
+            ..Default::default()
+        };
+
+        let egui_renderer = {
+            ash_egui::Renderer::new(
+                window_width,
+                window_height,
+                window_scale_factor,
+                &device.raw,
+                &device.physical_device().properties,
+                &device.physical_device().memory_properties,
+                context,
+                raw_input.clone(),
+            )
+        };
+
+        Self {
+            device,
+            inner: Arc::new(Mutex::new(EguiBackendInner {
+                egui_renderer,
+                gfx: None,
+            })),
+            raw_input,
+        }
+    }
+
+    pub fn create_graphics_resources(&mut self, surface_resolution: [u32; 2]) {
+        self.inner
+            .lock()
+            .create_graphics_resources(self.device.as_ref(), surface_resolution);
+    }
+
+    #[allow(dead_code)]
+    pub fn destroy_graphics_resources(&mut self) {
+        let device = &self.device.raw;
+
+        log::trace!("device_wait_idle");
+        unsafe { device.device_wait_idle() }.unwrap();
+
+        let mut inner = self.inner.lock();
+
+        if inner.egui_renderer.has_pipeline() {
+            inner.egui_renderer.destroy_pipeline(device);
+        }
+
+        if let Some(gfx) = inner.gfx.take() {
+            unsafe {
+                // TODO
+                //device.destroy_render_pass(gfx.egui_render_pass, None);
+                device.destroy_framebuffer(gfx.egui_framebuffer, None);
+            }
+        }
+    }
+
+    pub fn handle_event(
+        &mut self,
+        _window: &winit::window::Window,
+        _egui: &mut ash_egui::egui::Context,
+        _event: &winit::event::Event<'_, ()>,
+    ) {
+    }
+
+    pub fn prepare_frame(&mut self, context: &mut CtxRef, dt: f32) {
+        // update time
+        if let Some(time) = self.raw_input.time {
+            self.raw_input.time = Some(time + dt as f64);
+        } else {
+            self.raw_input.time = Some(0.0);
+        }
+
+        context.begin_frame(self.raw_input.take());
+    }
```

Comment:
I changed the frame prep API to this `EguiBackend::prepare_frame(&mut self.egui);` and you may wonder why this is no longer EguiBackend object method rather a static function, and that is because `prepare_frame()` only operates on fields of `EguiState`, while it has no need for data in `EguiBackend`.  

Basically, `EguiState` contains the egui `Context` instance and `RawInput` input state that the user of the library will interact with directly, while `EguiBackend` needs the data passed from `EguiState` once a frame to render.